### DISCUSSION TITLE
Command Console unable to Command (only reads) - UE 5.5

### DISCUSSION THREAD

**Discussion by SLGY** [JSBSim-Team/jsbsim#1210] [Q&A] [answered]
As per Unreal Engine section ReadMe, I'm implementing use of the Command Console (instead of having hardcode every additional input/output parameter I need into the UE plugin itself):

![image](https://github.com/user-attachments/assets/c2e415cf-56b6-43ee-86e4-2a1069cc302f)

But the Command Console doesn't seem to be having any effect on the flight model in game. I'm able to _input_ and _read_ every FCS command, but it simply doesn't do anything to the aircraft flight model when running.

So I went back and redownloaded reference app, compiled+ran it (I'm using UE 5.5, overcame the toolset issues etc, and have it running perfectly - until I make these below edits with Command Console). Here's the _default_ flight control node setup (which runs each Event Tick):
![Screenshot (53)](https://github.com/user-attachments/assets/f1574ad8-7c15-4276-aa3f-5eacc46b3314)

Now, for example, I remove the Set->Commands nodes and start with a Command Console node just to do Aileron command input (with _fcs/aileron-cmd-norm_ - which is read and write) I input the same normalised signal from before and have a print node to see the live output from the Command Console.
![Screenshot (54)](https://github.com/user-attachments/assets/e0a66962-7ec2-40a4-be86-8966fe154a17)

The result is that the aileron doesn't move (both animation and the flight model) but the output from the Command Console still gives the correct aileron demand that I input with the game pad? Note: it's not just an animation issue, I've done this with all the other inputs connected to the old Set->Command function and the inputs work with the Set->Commands node setup, but nothing works with the Command Console.
![Screenshot (55)](https://github.com/user-attachments/assets/6d30077a-069e-4486-974b-a3d9173ca8d1)

Am I missing something in how the Command Console function is intended to operate?

**Comment by gallonmate** [comment, marked as answer]
The Reference App has the main aircraft controls hardcoded and is being updated every simulation step using the variables directly from `Commands`, as seen in the Blueprint. `Commands`  is being copied via c++, regardless of the blueprint code, immediately before every simulation step, so it overrides any values you set outside of using the `Commands` variable.

You can use the `Commands` variable that already exists in the blueprint and update those values as needed. There's no need to use the Command Console Node for those controls. Although if for some reason you do not wish to use the `Commands` structure variable for the main controls, then you can edit the plugin in c++, by commenting out/deleting the controls you wish to override. Below is a section of code that sends the `Commands` values to JSBSim. The first one is for the Ailerons:

https://github.com/JSBSim-Team/jsbsim/blob/318c80408b4566addf42983a764362edbddbb792/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L593-L605

**Reply by SLGY** [reply]
Thanks, that's explained everything now. It seems the intent of making the Command Console was only to access inputs/outputs that aren't _already_ in the Commands struct. After I hardcoded a number of different parameters to make a relatively complex aircraft HUD, I saw that the Command Console (mentioned in the ReadMe) would save me having to do that each time - but I made the mistake of making a new aircraft from scratch based _solely_ on the Command Console.
Anyways, now I know what options I have moving forward. Thanks again.
Is it worth tagging the relevant person to update the UE readme section to make this clearer maybe?

**Reply by gallonmate** [reply]
Sorry for the confusion. I was actually the one who added the Command Console. I'm just a normal user/community member. The original plugin by the Epic Team did not have this function. When I was trying to get JSBSim Helicopters working, it proved impossible following the first method of hardcoding controls, because the helicopter properties are not defined yet in JSBSim, they are created at runtime as new properties. So as I dug into JSBSim's codebase to see how it worked under the hood in creating it's new JSBSim properties, I realized how powerful the JSBSim Property Manger is for accessing any property. And I quickly added it to the plugin with the basic intention of "Here's a new powerful function!"

I realize now that it is misleading in my original statement of it's ability to "get and set any command". So we can definitely change that and make it clearer.

I have other small updates to keep the plugin up-to-date that I will be submitting within the next week. I will re-word the Command Console section to be clearer. I can tag you if you'd like to be part of the review process when I submit the pull request.

**Reply by SLGY** [reply]
Thanks, yeah tag me for sure 👌

**Reply by gallonmate** [reply]
@SLGY Hi, I updated the ReadMe if you would like to review it. I tried to tag you directly in the pull request but it did not allow me to tag you there. You can find it in the Pulll Request tab or see the ReadMe changes here: https://github.com/JSBSim-Team/jsbsim/pull/1221/files#diff-15f8e17eb6f0b352b1cce3846b8852f4b6be879a77c5957db109d611a2a650ef

**Reply by SLGY** [reply]
Yeah that looks good, issue resolved. And I noticed the landing gear bug fix in there too, I just always thought there was something wrong with my blueprint setup and couldn't get it to spawn with the gear up. I should really visit here more often when I have these small issues.

I'll close the discussion now as it's resolved?

**Reply by gallonmate** [reply]
Thanks for reviewing the ReadMe! Sure you can close the discussion, although I think it's already marked closed/answered.

Also some info regarding the updated bug fix for gear down control. I didn't want to change how the original plugin was setup, so this is the way it works after the bug fix:

<ins>`GearDown`</ins> (in the Commands details) = Manual full range of the gear, anywhere from 0% to 100% Down (0.0 - 1.0).  

<ins>`StartwithGearDown`</ins>:
Enabled = Force the gear down on start. Forces  `GearDown` position to 1.
Disabled = Do nothing. Plugin uses the value set in `GearDown` position.


So to have gear up on spawn, StartwithGeardown needs to be disabled and Geardown needs to be 0.0.
